lib/listener/helpers.js

```diff
@@ -29,11 +29,27 @@ module.exports = function () {
   });
 
   event.dispatcher.on(event.suite.before, function (suite) {
-    runAsyncHelpersHook('_beforeSuite', suite, true);
+    try {
+      runAsyncHelpersHook('_beforeSuite', suite, true);
+    } catch (err) {
+      recorder.throw(err);
+    }
+    recorder.catch((err) => {
+      event.emit(event.test.failed, suite, err); // emit
+      throw err;
+    });
```

Comment:
code duplication. Should be wrapped into another function and reused for suite.before, suite.after, test.before